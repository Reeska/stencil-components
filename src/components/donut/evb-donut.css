:host {
  display: inline-flex;
  position: relative;
  font-size: 2em;
}
:host(.progress--with-text)::after {
  content: attr(data-progress) '%';
  display: block;
  font-size: 1em;
  left: 50%;
  position: absolute;
  top: 50%;
  transform: translate(-50%, -50%);
}
.progress {
  height: auto;
  position: relative;
  transform: rotate(-90deg);
  width: auto;
}

.progress__meter, .progress__value {
  fill: none;
}

.progress__meter {
  stroke: var(--evb-border-light);
}

.progress__value {
  stroke: var(--evb-primary-color);
}

.progress__value.arc-left {
  transform: scaleY(-1);
  transform-origin: center;
}
