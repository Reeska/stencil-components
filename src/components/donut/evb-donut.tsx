import { Component, ComponentInterface, Prop } from '@stencil/core';

const MIN_DIAMETER = 1;
const PADDING = 4;
const MAX_PERCENTAGE = 100;

@Component({
  tag: 'evb-donut',
  shadow: true,
  styleUrl: 'evb-donut.css'
})
export class EvbDonut implements ComponentInterface {

  @Prop() progress = 0;
  @Prop() invert = false;
  @Prop() diameter = 120;
  @Prop() thickness = 12;
  @Prop() text = false;

  /** the circumference fo the donut */
  private circumference = (this.diameter - this.thickness) * Math.PI;

  /**
   * Get the SVG viewbox value for the recalculated donut
   */
  get viewbox() {
    return `0 0 ${this.diameter} ${this.diameter}`;
  }

  get x() {
    return Math.max((this.diameter / 2 - this.thickness / 2), 0);
  }

  get fontSize() {
    return Math.max(MIN_DIAMETER, (this.diameter - PADDING - (this.thickness * 2)) / 2);
  }

  hostData() {
    return {
      class: {
        'progress--with-text': this.text
      },
      style: {
        fontSize: `${this.fontSize}px`
      },
      'data-progress': this.progress || 0
    };
  }

  render() {
    this.circumference = (this.diameter - this.thickness) * Math.PI;
    const progressValueClass = this.invert ? 'arc-left progress__value' : 'progress__value';
    const progressStyles = this.calculateProgressStyles(this.progress);

    return (
      <svg class='progress'
        width={this.diameter}
        height={this.diameter}
        viewBox={this.viewbox}>

        <circle class='progress__meter'
          cx={this.diameter / 2}
          cy={this.diameter / 2}
          r={this.x}
          stroke-width={this.thickness} />

        <circle
          class={progressValueClass}
          style={progressStyles}
          cx={this.diameter / 2}
          cy={this.diameter / 2}
          r={this.x}
          stroke-width={this.thickness} />
      </svg>
    );
  }

  /**
   * Calculate the progress and set styling properties accordingly
   */
  private calculateProgressStyles(value: number) {
    const progress = value / MAX_PERCENTAGE;
    const dashoffset = this.circumference * (1 - progress);
    return {
      strokeDashoffset: dashoffset,
      strokeDasharray: this.circumference
    };
  }
}
