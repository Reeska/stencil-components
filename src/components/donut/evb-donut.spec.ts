import { EvbDonut } from './evb-donut';

describe('EvbDonut', () => {
  let component: EvbDonut;

  beforeEach(() => component = new EvbDonut());

  it('should create', () => {
    expect(component).toBeTruthy();
    expect(component.render()).toMatchSnapshot();
  });
});
