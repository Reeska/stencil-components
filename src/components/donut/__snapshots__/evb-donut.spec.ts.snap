// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EvbDonut should create 1`] = `
Object {
  "$attrs$": Object {
    "class": "",
    "data-progress": 0,
    "style": Object {
      "fontSize": "46px",
    },
  },
  "$children$": Array [
    Object {
      "$attrs$": Object {
        "class": "progress",
        "height": 120,
        "viewBox": "0 0 120 120",
        "width": 120,
      },
      "$children$": Array [
        Object {
          "$attrs$": Object {
            "class": "progress__meter",
            "cx": 60,
            "cy": 60,
            "r": 54,
            "stroke-width": 12,
          },
          "$children$": null,
          "$elm$": undefined,
          "$flags$": 0,
          "$key$": undefined,
          "$name$": undefined,
          "$tag$": "circle",
        },
        Object {
          "$attrs$": Object {
            "class": "progress__value",
            "cx": 60,
            "cy": 60,
            "r": 54,
            "stroke-width": 12,
            "style": Object {
              "strokeDasharray": "339.29200658769764",
              "strokeDashoffset": "339.29200658769764",
            },
          },
          "$children$": null,
          "$elm$": undefined,
          "$flags$": 0,
          "$key$": undefined,
          "$name$": undefined,
          "$tag$": "circle",
        },
      ],
      "$elm$": undefined,
      "$flags$": 0,
      "$key$": undefined,
      "$name$": undefined,
      "$tag$": "svg",
    },
  ],
  "$elm$": undefined,
  "$flags$": 0,
  "$key$": undefined,
  "$name$": undefined,
  "$tag$": Object {},
}
`;
